module/zfs/abd.c

```diff
@@ -153,14 +167,37 @@ static abd_stats_t abd_stats = {
 	 */
 	{ "scatter_chunk_waste",		KSTAT_DATA_UINT64 },
 	/*
-	 * The number of linear ABDs which are currently allocated, excluding
-	 * ABDs which don't own their data (for instance the ones which were
-	 * allocated through abd_get_offset() and abd_get_from_buf()). If an
-	 * ABD takes ownership of its buf then it will become tracked.
+	 * The number of compound allocations of a given order.  These
+	 * allocations are spread over all currently allocated ABDs, and
+	 * act as a measure of memory fragmentation.
 	 */
-	{ "linear_cnt",				KSTAT_DATA_UINT64 },
-	/* Amount of data stored in all linear ABDs tracked by linear_cnt */
-	{ "linear_data_size",			KSTAT_DATA_UINT64 },
+	{ { "scatter_order_N",			KSTAT_DATA_UINT64 } },
+	/*
+	 * The number of scatter ABDs which contain multiple chunks.
+	 * ABDs are preferentially allocated from the minimum number of
+	 * contiguous multi-page chunks, a single chunk is optimal.
+	 */
+	{ "scatter_page_multi_chunk",		KSTAT_DATA_UINT64 },
+	/*
+	 * The number of scatter ABDs which are split accross memory zones.
+	 * ABDs are preferentially allocated using pages from a single zone.
+	 */
+	{ "scatter_page_split_zone",		KSTAT_DATA_UINT64 },
```

Comment:
Why do we care about NUMA placement? Vast majority of these buffers will sit in the ARC, right? From there, they'll be accessed by random processes, defeating any NUMA strategies we do here. Instead, doing interleaved allocation might give the best average performance (like pagecache allocations), and memory utilization.
Only operations that touch the data in zio pipeline would have some benefit, but only if the abd is dispatched to the correct thread. It might be more beneficial to allocate IO ABDs on NUMA node where the storage device is attached.
Am I missing something obvious?